core/services/keystore/keys/ocr2key/onchain_keyring.go

```diff
@@ -0,0 +1,90 @@
+package ocr2key
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"fmt"
+	"io"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/minio/sha256-simd"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/keytype"
+	"github.com/smartcontractkit/libocr/offchainreporting2/chains/evmutil"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2/types"
+)
+
+var _ ocrtypes.OnchainKeyring = &OnchainKeyring{}
+
+type OnchainKeyring struct {
+	privateKey ecdsa.PrivateKey
+	keyType    keytype.KeyType
+}
+
+func newOnchainKeyring(material io.Reader, keyType keytype.KeyType) (*OnchainKeyring, error) {
+	ecdsaKey, err := ecdsa.GenerateKey(curve, material)
+	if err != nil {
+		return nil, err
+	}
+	return &OnchainKeyring{privateKey: *ecdsaKey, keyType: keyType}, nil
+}
+
+// XXX: PublicKey returns the address of the public key not the public key itself
+func (ok *OnchainKeyring) PublicKey() ocrtypes.OnchainPublicKey {
+	address := ok.SigningAddress()
+	return address[:]
+}
+
+func (ok *OnchainKeyring) reportToSigData(reportCtx ocrtypes.ReportContext, report ocrtypes.Report) []byte {
+	rawReportContext := evmutil.RawReportContext(reportCtx)
+	sigData := ok.hash(report)
```

Comment:
ah one subtle difference - in the solana contract we use sha256(report || reportContext...) as opposed to evm ones which is keccak(keccak(report) || reportContext..). I think instead of trying to share the OnchainKeyring implementation between the two, I think it be cleaner to just defined an interface which embeds libocr's OnchainKeyring interface and just adds the SigningAddress + Marshal/Unmarshal functions that we need for the keystore. Then just have separate implementations and constructors for it i.e. NewSolanaKeyBundle and NewEthereumKeyBundle which can re-use the same offchain keyring instantiation.